srml/generic-asset/src/lib.rs

```diff
@@ -0,0 +1,1241 @@
+//! - `slash`: Deduct up to an amount from the combined balance of `who`, preferring to deduct from the
+//!	free balance. This function cannot fail.
+//! - `reward`: Add up to an amount to the free balance of an account.
+//! - `slash_reserved`: Deduct up to an amount from reserved balance of an account. This function cannot fail.
+//! - `repatriate_reserved`: Move up to an amount from reserved balance of an account to free balance of another
+//! account.
+//! - `check_permission`: Check permission to perform burn, mint or update.
+//! - `ensure_can_withdraw`: Check if the account is able to make a withdrawal of the given amount
+//!	for the given reason.
+//!
+//! ### Usage
+//!
+//! The following examples show how to use the Generic Asset module in your custom module.
+//!
+//! ### Examples from the PRML
+//!
+//! The Fees module uses the `Currency` trait to handle fee charge/refund, and its types inherit from `Currency`:
+//!
+//! ```
+//! use support::{
+//! 	traits::{Currency, ExistenceRequirement, WithdrawReason},
+//! 	dispatch::Result,
+//! };
+//! # pub trait Trait: system::Trait {
+//! # 	type Currency: Currency<Self::AccountId>;
+//! # }
+//! type AssetOf<T> = <<T as Trait>::Currency as Currency<<T as system::Trait>::AccountId>>::Balance;
+//!
+//! fn charge_fee<T: Trait>(transactor: &T::AccountId, amount: AssetOf<T>) -> Result {
+//! 	// ...
+//! 	T::Currency::withdraw(
+//! 		transactor,
+//! 		amount,
+//! 		WithdrawReason::TransactionPayment,
+//! 		ExistenceRequirement::KeepAlive,
+//! 	)?;
+//! 	// ...
+//! 	Ok(())
+//! }
+//!
+//! fn refund_fee<T: Trait>(transactor: &T::AccountId, amount: AssetOf<T>) -> Result {
+//! 	// ...
+//! 	T::Currency::deposit_into_existing(transactor, amount)?;
+//! 	// ...
+//! 	Ok(())
+//! }
+//!
+//! # fn main() {}
+//! ```
+//!
+//! ## Genesis config
+//!
+//! The Generic Asset module depends on the [`GenesisConfig`](./struct.GenesisConfig.html).
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use parity_codec::{Decode, Encode, HasCompact};
+
+use primitives::traits::{
+	CheckedAdd, CheckedSub, MaybeSerializeDebug, Member, One, Saturating, SimpleArithmetic, Zero,
+};
+
+use rstd::prelude::*;
+use rstd::{cmp, result};
+use support::dispatch::Result;
+use support::{
+	decl_event, decl_module, decl_storage, ensure,
+	traits::{
+		Currency, ExistenceRequirement, Imbalance, LockIdentifier, LockableCurrency, ReservableCurrency,
+		SignedImbalance, UpdateBalanceOutcome, WithdrawReason, WithdrawReasons,
+	},
+	Parameter, StorageDoubleMap, StorageMap, StorageValue,
+};
+use system::ensure_signed;
+
+mod mock;
+mod tests;
+
+pub use self::imbalances::{NegativeImbalance, PositiveImbalance};
+
+pub trait Trait: system::Trait {
+	type Balance: Parameter
+		+ Member
+		+ SimpleArithmetic
+		+ Default
+		+ Copy
+		+ MaybeSerializeDebug;
+	type AssetId: Parameter + Member + SimpleArithmetic + Default + Copy;
```

Comment:
i don't think this needs to be generic at all, and it complicates things a bit. would remove and just have a `pub type AssetId = u32;` declaration.

Reply:
It is not that complicated and I can think some reason why people may want to use other values instead of `u32`.
For a chain that only need to support fixed amount of tokens (e.g. 5), it is may not idea to have 5 `Balances` module instances. It will make sense to use this module with `u8` asset id or maybe `enum AssetId { Token1, Token2, Token3, Token4, Token5 }`

Reply:
Allowing representation with something like an `enum` breaks the idea that it's a module for _fungible_ assets. IMO if you want to represent something like tracking individual tokens as variants of an enum, that should be in a different module for non-fungible assets.

Reply:
FWIW: I accepted a pr for the `srml-assets` module to allow the `AssetId` to be generic: https://github.com/paritytech/substrate/pull/2838

Reply:
@joepetrowski Let me clarify it. Every token for a same asset id are fungible. Tokens between different asset ID are not fungible just like you can't mix two different ERC20 tokens.
In my example, a chain may support 5 different tokens: WrappedDOT, WrappedETH, WrappedBTC, WrappedUSD, WrappedEUR and `AssetId` can be typed of `enum Asset { WrappedDOT, WrappedETH, WrappedBTC, WrappedUSD, WrappedEUR }`. There are 5 different fungible assets.

Reply:
Sorry, I read your first comment as having an asset with very small number of tokens, each as a variant.

But wouldn't this enum live in the runtime and not the asset module?

Reply:
Yes. It is just a quick example and may not be practical.